src/hotspot/share/runtime/handshake.cpp

```diff
@@ -565,39 +567,31 @@ HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* ma
 
   Thread* current_thread = Thread::current();
 
-  HandshakeState::ProcessResult pr_ret = HandshakeState::_processed;
-  int executed = 0;
+  HandshakeOperation* op = get_op();
 
-  do {
-    HandshakeOperation* op = pop();
-    if (op != NULL) {
-      assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
-      assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
-      log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s(%s)", p2i(op),
-                           op == match_op ? "handshaker" : "cooperative",
-                           current_thread->is_VM_thread() ? "VM Thread" : "JavaThread");
-
-      if (op == match_op) {
-        pr_ret = HandshakeState::_succeeded;
-      }
+  assert(op != NULL, "Must have an op");
+  assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
+  assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
 
-      op->prepare(_handshakee, current_thread);
+  log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s(%s)", p2i(op),
+                       op == match_op ? "handshaker" : "cooperative",
+                       current_thread->is_VM_thread() ? "VM Thread" : "JavaThread");
 
-      _active_handshaker = current_thread;
-      op->do_handshake(_handshakee);
-      _active_handshaker = NULL;
+  op->prepare(_handshakee, current_thread);
 
-      executed++;
-    }
-  } while (have_non_self_executable_operation());
+  set_active_handshaker(current_thread);
+  op->do_handshake(_handshakee); // acquire, op removed after
+  set_active_handshaker(NULL);
+  remove_op(op);
 
   _lock.unlock();
 
-  log_trace(handshake)("%s(" INTPTR_FORMAT ") executed %d ops for JavaThread: " INTPTR_FORMAT " %s target op: " INTPTR_FORMAT,
+  log_trace(handshake)("%s(" INTPTR_FORMAT ") executed an op for JavaThread: " INTPTR_FORMAT " %s target op: " INTPTR_FORMAT,
                        current_thread->is_VM_thread() ? "VM Thread" : "JavaThread",
-                       p2i(current_thread), executed, p2i(_handshakee),
-                       pr_ret == HandshakeState::_succeeded ? "including" : "excluding", p2i(match_op));
-  return pr_ret;
+                       p2i(current_thread), p2i(_handshakee),
+                       op == match_op ? "including" : "excluding", p2i(match_op));
+
+  return op == match_op ? HandshakeState::_succeeded : HandshakeState::_processed;
```

Comment:
Pre-existing: what does _processed mean versus _succeeded?

Reply:
- Processed means we did progress, so do not yield.
- Succeeded means you processed the OP you add to the queue. (For your purposes you don't need to revisit this handshakee)